cmd/snap-update-ns/trespassing.go

```diff
@@ -0,0 +1,234 @@
+//
+// Assumptions constitute the global part of the write restriction management.
+// Assumptions are global in the sense that they span multiple distinct write
+// operations. In contrast, Restrictions track per-operation state.
+type Assumptions struct {
+	unrestrictedPaths []string
+	pastChanges       []*Change
+}
+
+// AddUnrestrictedPaths adds a list of directories where writing is allowed
+// even if it would hit the real host filesystem (or transit through the host
+// filesystem). This is intended to be used with certain well-known locations
+// such as /tmp, $SNAP_DATA and $SNAP.
+func (as *Assumptions) AddUnrestrictedPaths(paths ...string) {
+	as.unrestrictedPaths = append(as.unrestrictedPaths, paths...)
+}
+
+// isRestricted returns true if a path follows restricted writing scheme.
+//
+// Writing to a restricted path results in step-by-step validation of each
+// directory, starting from the root of the file system. Unless writing is
+// allowed a mimic must be constructed to ensure that writes are not visible in
+// undesired locations of the host filesystem.
+//
+// Provided path is the full, absolute path of the entity that needs to be
+// created (directory, file or symbolic link).
+func (as *Assumptions) isRestricted(path string) bool {
+	// Anything rooted at one of the unrestricted paths is not restricted.
+	// Those are for things like /var/snap/, for example.
+	for _, p := range as.unrestrictedPaths {
+		if p == "/" || p == path || strings.HasPrefix(path, filepath.Clean(p)+"/") {
+			return false
+		}
+
+	}
+	// All other paths are restricted
+	return true
+}
+
+// MockUnrestrictedPaths replaces the set of path paths without any restrictions.
+func (as *Assumptions) MockUnrestrictedPaths(paths ...string) (restore func()) {
+	old := as.unrestrictedPaths
+	as.unrestrictedPaths = paths
+	return func() {
+		as.unrestrictedPaths = old
+	}
+}
+
+// AddChange records the fact that a change was applied to the system.
+func (as *Assumptions) AddChange(change *Change) {
+	as.pastChanges = append(as.pastChanges, change)
+}
+
+// canWriteToDirectory returns true if writing to a given directory is allowed.
+//
+// Writing is allowed in one of thee cases:
+// 1) The directory is in one of the explicitly permitted locations.
+//    This is the strongest permission as it explicitly allows writing to
+//    places that may show up on the host, one of the examples being $SNAP_DATA.
+// 2) The directory is on a read-only filesystem.
+// 3) The directory is on a tmpfs created by snapd.
+func (as *Assumptions) canWriteToDirectory(dirFd int, dirName string) (bool, error) {
+	if !as.isRestricted(dirName) {
+		return true, nil
+	}
+	var fsData syscall.Statfs_t
+	if err := sysFstatfs(dirFd, &fsData); err != nil {
+		return false, fmt.Errorf("cannot fstatfs %q: %s", dirName, err)
+	}
+	// Writing to read only directories is allowed because EROFS is handled
+	// by each of the writing helpers already.
```

Comment:
This looks a little peculiar. Can you elaborate a little more (perhaps in the code, if it makes sens) what are the general semantics, i.e. why do we say 'true' if we know it's false?

Reply:
We are allowing _attempts_ to write because we know the code already handles EROFS correctly. We are going to say "yeah, go ahead, you cannot do any harm here". Note that we _attempt_ to write but we also handle EEXIST in the code so that when we are looking at creating `/usr/share/foo` we _attempt_ to create `/usr` but it exists so there is no error, then we _attempt_ to create `share` but it also exists then we attempt to create `foo` but it doesn't exist so we get EROFS which is handled already. If we said no here you would get a mimic even _before_ you considered if `usr` exists (because it is racy not to attempt that) and the whole thing would fall apart.

Reply:
Makes sense, thanks for explanation.